api/api-public-tenant-permissions.go

```diff
@@ -130,37 +103,164 @@ func (s *server) AddPermission(ctx context.Context, in *pb.AddPermissionRequest)
+	}
+	if len(actionTypes) == 0 {
+		return nil, status.New(codes.InvalidArgument, "a list of actions is needed").Err()
+	}
+	if permissionEffect == "" {
+		return nil, status.New(codes.InvalidArgument, "a valid effect is needed").Err()
+	}
+	if permissionName == "" {
+		return nil, status.New(codes.InvalidArgument, "a valid permission name  is needed").Err()
+	}
+	if description == "" {
+		return nil, status.New(codes.InvalidArgument, "a valid description is needed").Err()
+	}
+	effect := cluster.EffectFromString(permissionEffect)
+	if err := effect.IsValid(); err != nil {
+		return nil, status.New(codes.InvalidArgument, "invalid effect").Err()
+	}
+	// start app context
+	appCtx, err := cluster.NewTenantContextWithGrpcContext(ctx)
+	if err != nil {
+		return nil, err
+	}
 
-	for _, permResource := range permissionObj.Resources {
-		pbResource := pb.PermissionResource{
-			Id:         permResource.ID.String(),
-			BucketName: permResource.BucketName,
-			Pattern:    permResource.Pattern,
+	defer func() {
+		if err != nil {
+			appCtx.Rollback()
+			return
 		}
-		res.Resources = append(res.Resources, &pbResource)
+	}()
+
+	// get permission
+	permission, err := cluster.GetPermissionByID(appCtx, id)
+	if err != nil {
+		return nil, status.New(codes.InvalidArgument, "permission not found").Err()
 	}
-	for _, permAction := range permissionObj.Actions {
-		pbAction := pb.PermissionAction{
-			Id:   permAction.ID.String(),
-			Type: string(permAction.ActionType),
+
+	// start updating values on the permission obj not yet on db.
+	permission.Name = permissionName
+	permission.Effect = effect
+	permission.Description = &description
+
+	// Nullified values if they are empty
+	if description == "" {
+		permission.Description = nil
+	}
+
+	// -- Update Permission RESOURCES
+	var currentResourceBucketNames []string
+	mapResourceName := make(map[string]string)
+	for _, perm := range permission.Resources {
+		currentResourceBucketNames = append(currentResourceBucketNames, perm.BucketName)
+		mapResourceName[perm.BucketName] = perm.ID.String()
+	}
+	// TODO: parallelize
+	resourcesToCreate := differenceArrays(resourcesBucketNames, currentResourceBucketNames)
+	resourcesToDelete := differenceArrays(currentResourceBucketNames, resourcesBucketNames)
+	// CREATE New Resources
+	// create a Temporal permission to Create the new Permission resources
+	tempPermission := &cluster.Permission{ID: permission.ID}
+	cluster.AppendPermissionResourcesObj(tempPermission, resourcesToCreate)
+	// for each resource, save to DB
+	for _, resc := range tempPermission.Resources {
+		err = cluster.InsertResource(appCtx, tempPermission, &resc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// DELETE unwanted Resources
+	for _, bucketName := range resourcesToDelete {
+		resourceID, ok := mapResourceName[bucketName]
+		if !ok {
+			return nil, status.New(codes.Internal, "error retrieving resource to delete").Err()
+		}
+		err = cluster.DeletePermissionResourceDB(appCtx, resourceID)
```

Comment:
Do we support bulk delete instead of deleting the records one by one?

Reply:
Done